Readme.md
# retext-search [![Build Status](https://img.shields.io/travis/wooorm/retext-search.svg?style=flat)](https://travis-ci.org/wooorm/retext-search) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-search.svg?style=flat)](https://coveralls.io/r/wooorm/retext-search?branch=master)

Search in a document with **[Retext](https://github.com/wooorm/retext)**.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-search
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-search
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-search
```

[Duo](http://duojs.org/#getting-started):

```javascript
var search = require('wooorm/retext-search');
```

## Usage

```javascript
var Retext = require('retext');
var search = require('retext-search');
var inspect = require('retext-inspect');

var retext = new Retext()
    .use(inspect)
    .use(search);

retext.parse(
    'Clair wants to drink milk after her morning run. ' +
    'Schmidt likes coffee in the morning, when reading a book.' +

    '\n\n' +

    'Ms Clare likes her tea with milk, while sugar is prefered by ' +
    'Mr Smith. ' +
    'Xavier likes rowing in the morning.',
    function (err, tree) {
        console.log(tree.search('smit'));
        /*
         * [
         *   // The `WordNode` in the second sentence:
         *   WordNode[1]
         *   └─ TextNode: 'Schmidt',
         *
         *   // The `WordNode` in the penultimate sentence
         *   WordNode[1]
         *   └─ TextNode: 'Smith'
         * ]
         */
    }
);
```

## API

### [TextOM.Parent](https://github.com/wooorm/textom#textomparent-nlcstparent)#search(query)

Search the operated on context for `query`.

Parameters:

- `query` (`string|array`): sentence containing words to search for.

Returns: an array of [`WordNode`](https://github.com/wooorm/textom#textomwordnode-nlcstwordnode)s.

```javascript
tree.search('test');
// []

tree.search('xavier clair');
/*
 * [
 *   WordNode[1]
 *   └─ TextNode: 'Clair',
 *   WordNode[1]
 *   └─ TextNode: 'Clare',
 *   WordNode[1]
 *   └─ TextNode: 'Xavier'
 * ]
 */

tree.search(['milk']);
/*
 * [
 *   WordNode[1]
 *   └─ TextNode: 'milk',
 *   WordNode[1]
 *   └─ TextNode: 'milk'
 * ]
 */
```

### [TextOM.Parent](https://github.com/wooorm/textom#textomparent-nlcstparent)#searchAll(query)

Search the operated on context for `query`. Useful for ranking sentences or paragraphs.

Parameters:

- `query` (`string|array`): sentence containing words to search for.

Returns: an array of match objects. A match object contains a `node` property, and a `matches` array (containing match objects).

```javascript
tree.searchAll('test'); // null
tree.searchAll('xavier clair');
/*
 * Array
 * ├─ 0:
 * |  ├─ node: ParagraphNode
 * |  └─ matches:
 * |      └─ 0:
 * |         ├─ node: SentenceNode
 * |         └─ matches:
 * |             └─ 0: WordNode: "Clair"
 * └─ 1:
 *    ├─ node: ParagraphNode
 *    └─ matches:
 *        |─ 0:
 *        |  ├─ node: SentenceNode
 *        |  └─ matches:
 *        |      └─ 0: WordNode: "Clare"
 *        └─ 1:
 *           ├─ node: SentenceNode
 *           └─ matches:
 *               └─ 0: WordNode: "Xavier"
 */
```

## Benchmark

On a MacBook Air:

```text
             A section (10 paragraphs, 50 sentences, 300 words)
  4,560 op/s » Searching in lipsum for "Lorem"
  5,044 op/s » Searching parents in lipsum for "Lorem"

             An article (100 paragraphs, 500 sentences, 3,000 words)
    858 op/s » Searching in lipsum for "Lorem"
    823 op/s » Searching parents in lipsum for "Lorem"

             A book (1,000 paragraphs, 5,000 sentences, 30,000 words)
     88 op/s » Searching in lipsum for "Lorem"
     65 op/s » Searching parents in lipsum for "Lorem"
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.2.7 / 2015-01-18
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update retext
  * Update eslint

0.2.6 / 2014-12-14
==================

 * Update benchmark results in `Readme.md`
 * Fix wording in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.2.5 / 2014-11-22
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Add retext-inspect to dev-dependencies
 * Update retext, matcha

0.2.4 / 2014-10-28
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext, retext-content, retext-double-metaphone, retext-porter-stemmer, retext-visit

0.2.3 / 2014-10-21
==================

 * Refactor .jscs.json
 * Update retext

0.2.2 / 2014-10-15
==================

 * Update API for changes in retext-visit
 * Move `benchmark/` to `benchmark.js`
 * Update dependencies
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.2.1 / 2014-09-27
==================

 * Add missing comma in component.json

0.2.0 / 2014-09-26
==================

 * Refactor Readme.md
 * Refactor benchmark
 * Update .gitignore, .npmignore
 * Remove browser test
 * Add npm deployment to travis
 * Remove `npm update` from travis
 * Update Installation in docs
 * Add bower.json
 * Fix property order in component.json, package.json
 * Update copyright in Readme.md
 * Remove testling
 * Refactor spec
 * Refactor API
 * Add retext-content as a dependency
 * Update dependencies
 * Add components to gitignore

0.1.0 / 2014-07-13
==================

 * Added changelog
 * Fixed for API changes in retext
 * Removed parse-english only functionality
 * Fixed bug in benchmarks, update docs
 * Update retext-double-metaphone, retext-porter-stemmer, retext-visit, eslint, istanbul, retext
 * Removed functionality to browserify unit tests by default
 * Fix line ending

0.0.1 / 2014-06-25
==================


test.js
'use strict';

/*
 * Dependencies.
 */

var Retext,
    search,
    inspect,
    assert;

search = require('./');
Retext = require('retext');
inspect = require('retext-inspect');
assert = require('assert');

/*
 * Fixtures.
 */

var value;

value =
    'Clair wants to drink milk after her morning run. ' +
    'Schmidt likes coffee in the morning, when reading a book.\n' +
    '\n' +
    'Ms Clare likes her tea with milk, while sugar is prefered by ' +
    'Mr Smith. ' +
    'Xavier likes rowing in the morning.';

/*
 * Retext.
 */

var retext,
    TextOM;

retext = new Retext()
    .use(inspect)
    .use(search);

TextOM = retext.TextOM;

/*
 * Tests.
 */

describe('retext-search()', function () {
    it('should be a `function`', function () {
        assert(typeof search === 'function');
    });
});

describe('Parent#search(values)', function () {
    var tree;

    before(function (done) {
        retext.parse(value, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof TextOM.Parent.prototype.search === 'function');
    });

    it('should return an empty array when a falsey input is given',
        function () {
            assert(tree.search().length === 0);
            assert(tree.search([]).length === 0);
            assert(tree.search(null).length === 0);
        }
    );

    it('should throw when a valid input without words is given', function () {
        assert.throws(function () {
            tree.search('.');
        }, /`\.`/);

        assert.throws(function () {
            tree.search(':)');
        }, /`\:\)`/);
    });

    it('should return an empty array when nothing matches', function () {
        assert(tree.search('test').length === 0);
    });

    it('should return the matched words when given a string', function () {
        var matches;

        matches = tree.search('Xavier');

        assert(matches[0] === tree.tail.tail.head);
        assert(matches.length === 1);
    });

    it('should return the matched words when given an array', function () {
        var matches;

        matches = tree.search(['Xavier']);

        assert(matches[0] === tree.tail.tail.head);
        assert(matches.length === 1);
    });

    it('should return the matched words based on phonetics', function () {
        var matches;

        matches = tree.search('Smit');

        assert(matches[0] === tree.head.tail.head);
        assert(matches[1] === tree.tail.head.tail.prev);
        assert(matches.length === 2);
    });

    it('should return different matched words', function () {
        var matches;

        matches = tree.search('xavier clair');

        assert(matches[0] === tree.head.head.head);
        assert(matches[1] === tree.tail.head[2]);
        assert(matches[2] === tree.tail.tail.head);
        assert(matches.length === 3);
    });
});

describe('Parent#searchAll(values)', function () {
    var tree;

    before(function (done) {
        retext.parse(value, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof TextOM.Parent.prototype.searchAll === 'function');
    });

    it('should return an empty array when a falsey input is given',
        function () {
            assert(tree.searchAll().length === 0);
            assert(tree.searchAll([]).length === 0);
            assert(tree.searchAll(null).length === 0);
        }
    );

    it('should throw when a valid input without words is given', function () {
        assert.throws(function () {
            tree.searchAll('.');
        }, /`\.`/);

        assert.throws(function () {
            tree.searchAll(':)');
        }, /`\:\)`/);
    });

    it('should return an empty array when nothing matches', function () {
        assert(tree.searchAll('test').length === 0);
    });

    it('should return an object containing an array, containing each ' +
        'matched direct child', function () {
            var matches;

            matches = tree.searchAll('Xavier');

            assert(matches[0].node === tree.tail);
            assert(matches.length === 1);

            matches = tree.tail.tail.searchAll('Xavier');

            assert(matches[0] === tree.tail.tail.head);
            assert(matches.length === 1);
        }
    );

    it('should return the matched words based on phonetics', function () {
        var matches;

        matches = tree.searchAll('Smit');

        assert(matches[0].node === tree.head);
        assert(matches[0].matches[0].node === tree.head.tail);
        assert(matches[0].matches.length === 1);
        assert(matches[0].matches[0].matches[0] === tree.head.tail.head);
        assert(matches[0].matches[0].matches.length === 1);
        assert(matches[1].node === tree.tail);
        assert(matches[1].matches[0].node === tree.tail.head);
        assert(matches[1].matches.length === 1);
        assert(matches[1].matches[0].matches[0] === tree.tail.head.tail.prev);
        assert(matches[1].matches[0].matches.length === 1);
        assert(matches.length === 2);
    });

    it('should return different matched words', function () {
        var matches;

        matches = tree.searchAll('xavier clair');

        assert(matches[0].node === tree.head);
        assert(matches[0].matches[0].node === tree.head.head);
        assert(matches[0].matches.length === 1);
        assert(matches[0].matches[0].matches[0] === tree.head.head.head);
        assert(matches[0].matches[0].matches.length === 1);
        assert(matches[1].node === tree.tail);
        assert(matches[1].matches[0].node === tree.tail.head);
        assert(matches[1].matches[1].node === tree.tail.tail);
        assert(matches[1].matches.length === 2);
        assert(matches[1].matches[0].matches[0] === tree.tail.head[2]);
        assert(matches[1].matches[0].matches.length === 1);
        assert(matches[1].matches[1].matches[0] === tree.tail.tail.head);
        assert(matches[1].matches[1].matches.length === 1);
        assert(matches.length === 2);
    });
});


spec/retext-search.spec.js
'use strict';

/**
 * Dependencies.
 */

var search,
    Retext,
    assert;

search = require('..');
Retext = require('retext');
assert = require('assert');

/**
 * Fixtures.
 */

var value;

value =
    'Clair wants to drink milk after her morning run. ' +
    'Schmidt likes coffee in the morning, when reading a book.\n' +
    '\n' +
    'Ms Clare likes her tea with milk, while sugar is prefered by ' +
    'Mr Smith. ' +
    'Xavier likes rowing in the morning.';

/**
 * Retext.
 */

var retext,
    TextOM;

retext = new Retext().use(search);

TextOM = retext.TextOM;

/**
 * Tests.
 */

describe('retext-search()', function () {
    it('should be a `function`', function () {
        assert(typeof search === 'function');
    });

    it('should export an `attach` method', function () {
        assert(typeof search.attach === 'function');
    });
});

describe('Parent#search(values)', function () {
    var tree;

    before(function (done) {
        retext.parse(value, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof TextOM.Parent.prototype.search === 'function');
    });

    it('should return an empty array when a falsey input is given',
        function () {
            assert(tree.search().length === 0);
            assert(tree.search([]).length === 0);
            assert(tree.search(null).length === 0);
        }
    );

    it('should throw when a valid input without words is given', function () {
        assert.throws(function () {
            tree.search('.');
        }, /`\.`/);

        assert.throws(function () {
            tree.search(':)');
        }, /`\:\)`/);
    });

    it('should return an empty array when nothing matches', function () {
        assert(tree.search('test').length === 0);
    });

    it('should return the matched words when given a string', function () {
        var matches;

        matches = tree.search('Xavier');

        assert(matches[0] === tree.tail.tail.head);
        assert(matches.length === 1);
    });

    it('should return the matched words when given an array', function () {
        var matches;

        matches = tree.search(['Xavier']);

        assert(matches[0] === tree.tail.tail.head);
        assert(matches.length === 1);
    });

    it('should return the matched words based on phonetics', function () {
        var matches;

        matches = tree.search('Smit');

        assert(matches[0] === tree.head.tail.head);
        assert(matches[1] === tree.tail.head.tail.prev);
        assert(matches.length === 2);
    });

    it('should return different matched words', function () {
        var matches;

        matches = tree.search('xavier clair');

        assert(matches[0] === tree.head.head.head);
        assert(matches[1] === tree.tail.head[2]);
        assert(matches[2] === tree.tail.tail.head);
        assert(matches.length === 3);
    });
});

describe('Parent#searchAll(values)', function () {
    var tree;

    before(function (done) {
        retext.parse(value, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof TextOM.Parent.prototype.searchAll === 'function');
    });

    it('should return an empty array when a falsey input is given',
        function () {
            assert(tree.searchAll().length === 0);
            assert(tree.searchAll([]).length === 0);
            assert(tree.searchAll(null).length === 0);
        }
    );

    it('should throw when a valid input without words is given', function () {
        assert.throws(function () {
            tree.searchAll('.');
        }, /`\.`/);

        assert.throws(function () {
            tree.searchAll(':)');
        }, /`\:\)`/);
    });

    it('should return an empty array when nothing matches', function () {
        assert(tree.searchAll('test').length === 0);
    });

    it('should return an object containing an array, containing each ' +
        'matched direct child', function () {
            var matches;

            matches = tree.searchAll('Xavier');

            assert(matches[0].node === tree.tail);
            assert(matches.length === 1);

            matches = tree.tail.tail.searchAll('Xavier');

            assert(matches[0] === tree.tail.tail.head);
            assert(matches.length === 1);
        }
    );

    it('should return the matched words based on phonetics', function () {
        var matches;

        matches = tree.searchAll('Smit');

        assert(matches[0].node === tree.head);
        assert(matches[0].matches[0].node === tree.head.tail);
        assert(matches[0].matches.length === 1);
        assert(matches[0].matches[0].matches[0] === tree.head.tail.head);
        assert(matches[0].matches[0].matches.length === 1);
        assert(matches[1].node === tree.tail);
        assert(matches[1].matches[0].node === tree.tail.head);
        assert(matches[1].matches.length === 1);
        assert(matches[1].matches[0].matches[0] === tree.tail.head.tail.prev);
        assert(matches[1].matches[0].matches.length === 1);
        assert(matches.length === 2);
    });

    it('should return different matched words', function () {
        var matches;

        matches = tree.searchAll('xavier clair');

        assert(matches[0].node === tree.head);
        assert(matches[0].matches[0].node === tree.head.head);
        assert(matches[0].matches.length === 1);
        assert(matches[0].matches[0].matches[0] === tree.head.head.head);
        assert(matches[0].matches[0].matches.length === 1);
        assert(matches[1].node === tree.tail);
        assert(matches[1].matches[0].node === tree.tail.head);
        assert(matches[1].matches[1].node === tree.tail.tail);
        assert(matches[1].matches.length === 2);
        assert(matches[1].matches[0].matches[0] === tree.tail.head[2]);
        assert(matches[1].matches[0].matches.length === 1);
        assert(matches[1].matches[1].matches[0] === tree.tail.tail.head);
        assert(matches[1].matches[1].matches.length === 1);
        assert(matches.length === 2);
    });
});


index.js
'use strict';

/*
 * Dependencies.
 */

var stemmer,
    doubleMetaphone,
    visit,
    content;

stemmer = require('retext-porter-stemmer');
doubleMetaphone = require('retext-double-metaphone');
visit = require('retext-visit');
content = require('retext-content');

/**
 * Test if `node` matches a query.
 *
 * @param {Array.<{0: string, 1: string}>} queries
 * @param {WordNode} node
 * @return {boolean}
 */
function test(queries, node) {
    var index,
        phonetics,
        query;

    index = queries.length;
    phonetics = node.data.stemmedPhonetics;

    while (index--) {
        query = queries[index];

        if (
            phonetics[0] === query[0] ||
            phonetics[1] === query[0] ||
            phonetics[0] === query[1] ||
            phonetics[1] === query[1]
        ) {
            return true;
        }
    }

    return false;
}

/**
 * Find words in `node` matching `values`.
 *
 * @param {string} values
 * @this {Parent}
 * @return {Array.<WordNode>}
 */
function search(values) {
    var self,
        words,
        query,
        result;

    self = this;

    result = [];

    if (values) {
        values = String(values);
    }

    if (!values || !values.length) {
        return result;
    }

    query = [];

    /*
     * Get the stems from the query.
     */

    words = new self.TextOM.SentenceNode();
    words.replaceContent(values);

    words.visit(self.WORD_NODE, function (node) {
        query.push(node.data.stemmedPhonetics);
    });

    if (!query.length) {
        throw new TypeError(
            'TypeError: `' + values + '` is not a valid argument ' +
            'for Parent#search(values)'
        );
    }

    /*
     * Find the query in `self`.
     */

    self.visit(self.WORD_NODE, function (node) {
        if (test(query, node)) {
            result.push(node);
        }
    });

    return result;
}

/**
 * Find the parents belonging to the nodes in
 * `matches`.
 *
 * @param {Array.<Node>} matches
 * @return {Object.<{matches: Array.<Node>, node: Parent}>}
 */
function flattenParents(matches) {
    var parents,
        scores,
        parentToScoreMap,
        index,
        length,
        child,
        parent,
        position,
        score;

    parents = [];
    scores = [];
    parentToScoreMap = [];

    index = -1;
    length = matches.length;

    while (++index < length) {
        child = matches[index];

        parent = (child.node || child).parent;

        position = parents.indexOf(parent);

        parents[index] = parent;

        if (position === -1) {
            position = scores.length;

            score = scores[position] = {
                'matches': [],
                'node': parent
            };
        } else {
            score = scores[parentToScoreMap[position]];
        }

        parentToScoreMap[index] = position;

        score.matches.push(child);
    }

    return scores;
}

/**
 * Find parents in `node` matching `values`.
 *
 * @param {string} values
 * @this {Parent}
 * @return {Array.<{node: Node, matches: Array}>}
 */
function searchAll(values) {
    var self,
        nodes;

    self = this;
    nodes = self.search(values);

    if (!nodes || !nodes.length) {
        return nodes;
    }

    /*
     * Keep op flattening the results object until
     * self is reached.
     */

    while (nodes[0].node !== self) {
        nodes = flattenParents(nodes);
    }

    /*
     * Return the matches of `self`.
     */

    return nodes[0].matches;
}

/**
 * Define `retextSearch`.
 *
 * @param {Retext} retext
 */
function retextSearch(retext) {
    var TextOM,
        parentPrototype,
        elementPrototype;

    TextOM = retext.TextOM;
    parentPrototype = TextOM.Parent.prototype;
    elementPrototype = TextOM.Element.prototype;

    retext
        .use(doubleMetaphone)
        .use(content)
        .use(stemmer)
        .use(visit);

    parentPrototype.search = elementPrototype.search = search;
    parentPrototype.searchAll = elementPrototype.searchAll = searchAll;
}

/*
 * Expose `retextSearch`.
 */

module.exports = retextSearch;


benchmark.js
'use strict';

/*
 * Dependencies.
 */

var search,
    Retext;

search = require('./');
Retext = require('retext');

/*
 * Fixtures.
 */

var tiny,
    small,
    medium,
    large;

tiny = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam ' +
    'ac ultricies diam, quis vehicula mauris. Vivamus accumsan eleifend ' +
    'quam et varius. Etiam congue id magna eu fermentum. Aliquam mollis ' +
    'adipiscing.\n\n';

small = Array(11).join(tiny);
medium = Array(11).join(small);
large = Array(11).join(medium);

/*
 * Retext.
 */

var retext;

retext = new Retext().use(search);

/*
 * Benchmarks.
 */

suite('A section (10 paragraphs, 50 sentences, 300 words)', function () {
    var tree;

    before(function (next) {
        retext.parse(small, function (err, node) {
            if (err) {
                throw err;
            }

            tree = node;
            next();
        });
    });

    bench('Searching in lipsum for "Lorem"', function () {
        tree.search('lorem');
    });

    bench('Searching parents in lipsum for "Lorem"', function () {
        tree.searchAll('lorem');
    });
});

suite('An article (100 paragraphs, 500 sentences, 3,000 words)', function () {
    var tree;

    before(function (next) {
        retext.parse(medium, function (err, node) {
            if (err) {
                throw err;
            }

            tree = node;
            next();
        });
    });

    bench('Searching in lipsum for "Lorem"', function () {
        tree.search('lorem');
    });

    bench('Searching parents in lipsum for "Lorem"', function () {
        tree.searchAll('lorem');
    });
});

suite('A book (1,000 paragraphs, 5,000 sentences, 30,000 words)',
    function () {
        var tree;

        before(function (next) {
            retext.parse(large, function (err, node) {
                if (err) {
                    throw err;
                }

                tree = node;
                next();
            });
        });

        bench('Searching in lipsum for "Lorem"', function () {
            tree.search('lorem');
        });

        bench('Searching parents in lipsum for "Lorem"', function () {
            tree.searchAll('lorem');
        });
    }
);
